--- vi.c
+++ vi.c
@@ -226,17 +226,38 @@
 static int vi_buf[128];
 static int vi_buflen;
 
-static int vi_read(void)
-{
-	return vi_buflen ? vi_buf[--vi_buflen] : term_read();
-}
-
 static void vi_back(int c)
 {
 	if (vi_buflen < sizeof(vi_buf))
 		vi_buf[vi_buflen++] = c;
 }
 
+static int vi_read(void)
+{
+	int c = vi_buflen ? vi_buf[--vi_buflen] : term_read();
+	
+	if (c == TK_ESC) {
+		int c2 = term_read();
+		if (c2 == '[') {
+			int c3 = term_read();
+			switch (c3) {
+			case 'A': return 'k';
+			case 'B': return 'j';
+			case 'C': return 'l';
+			case 'D': return 'h';
+			default:
+				vi_back(c3);
+				vi_back('[');
+				break;
+			}
+		} else if (c2 != -1) {
+			vi_back(c2);
+		}
+	}
+	
+	return c;
+}
+
 static char *vi_char(void)
 {
 	return led_read(&xkmap);
